protocol/bin-prot/src/loose_deserializer.rs

```diff
@@ -0,0 +1,193 @@
+                    BinProtRule::Bool => visitor.visit_bool(self.rdr.bin_read_bool()?),
+                    BinProtRule::Sum(summands) => {
+                        // read the enum variant index.
+                        // We need this to select which variant layout to use
+                        // when deserializing the variants data
+                        let index = self.rdr.bin_read_variant_index()?;
+                        let variant_rules = summands[index as usize].ctor_args.clone();
+                        self.mode
+                            .layout_iter
+                            .push(vec![BinProtRule::Tuple(variant_rules)]);
+                        visitor.visit_enum(ValueEnum::new(self, summands[index as usize].clone()))
+                    }
+                    BinProtRule::Option(some_rule) => {
+                        let index = self.rdr.bin_read_variant_index()?; // 0 or 1
+                        match index {
+                            0 => visitor.visit_none(),
+                            1 => {
+                                self.mode.layout_iter.push(vec![*some_rule]);
+                                visitor.visit_some(self)
+                            }
+                            _ => Err(Error::InvalidOptionByte { got: index }),
+                        }
+                    }
+                    BinProtRule::String => visitor.visit_bytes(&self.rdr.bin_read_bytes()?),
+                    BinProtRule::Float => visitor.visit_f64(self.rdr.read_f64::<LittleEndian>()?),
+                    BinProtRule::Char => {
+                        let c = self.rdr.read_u8()?;
+                        visitor.visit_char(c as char)
+                    }
+                    BinProtRule::List(element_rule) => {
+                        // read the length
+                        let len = self.rdr.bin_read_nat0()?;
+                        // request the iterator repeats the list elements the current number of times
+                        self.mode.layout_iter.push_n(*element_rule, len);
+                        // read the elements
+                        visitor.visit_seq(SeqAccess::new_list(self, len))
+                    }
+                    BinProtRule::Int
+                    | BinProtRule::Int32
+                    | BinProtRule::Int64
+                    | BinProtRule::NativeInt => visitor.visit_i64(self.rdr.bin_read_integer()?),
+                    BinProtRule::Polyvar(_)
+                    | BinProtRule::Vec(_, _)
+                    | BinProtRule::Nat0
+                    | BinProtRule::Hashtable(_)
+                    | BinProtRule::TypeVar(_)
+                    | BinProtRule::Bigstring
+                    | BinProtRule::SelfReference(_)
+                    | BinProtRule::TypeClosure(_, _)
+                    | BinProtRule::TypeAbstraction(_, _)
+                    | BinProtRule::Reference(_) => Err(Error::UnimplementedRule), // Don't know how to implement these yet
```

Comment:
I'm guessing these rules are never used in the layouts produced by Mina types and so not required. The loose deserialization is only used for testing so if we are able to parse the layouts and run our tests this seems fine to me.

Reply:
What happens if any of these rules become used in layouts produced by Mina, would CI fail somewhere so that is it known?  I think it's fine as long as it fails and we know this has happened.

Reply:
It will throw a nice `UnimplementedRule` error and probably a trace so we could figure out what is missing